Lift restrictions for matching of binaries and maps

bjorng opened the pull request:
Lift restrictions for matching of binaries and maps

There has always been an implementation limitation for matching
of binaries (for technical reasons). For example:

    foo(Bin) ->
        <<A:8>> = <<X:4,Y:4>> = Bin,
        {A,X,Y}.

This would fail to compile with the following message:

    t.erl:5:5: binary patterns cannot be matched in parallel using '='
    %    5|     <<A:8>> = <<X:4,Y:4>> = Bin,
    %     |     ^

This commit lifts this restriction, making the example legal.

A restriction for map matching is also lifted, but before we can
describe that, we'll need a digression to talk about the `=` operator.

The `=` operator can be used for two similar but slightly differently
purposes.

When used in a pattern in a clause, for example in a function head,
both the left-hand and right-hand side operands must be patterns:

    Pattern1 = Pattern2

For example:

    bar(#{a := A} = #{b := B}) -> {A, B}.

The following example will not compile because the right-hand side
is not a pattern but an expression:

    wrong(#{a := A} = #{b => B}) -> {A, B}.

    t.erl:4:23: illegal pattern
    %    4| wrong(#{a := A} = #{b => B}) -> {A, B}.
    %     |                       ^

Used in this context, the `=` operator does not imply that the two
patterns are matched in any particular order. Attempting to use a
variable matched out on the left-hand side on the right-hand side, or
vice versa, will fail:

    also_wrong1(#{B := A} = #{b := B}) -> {A,B}.
    also_wrong2(#{a := A} = #{A := B}) -> {A,B}.

    t.erl:6:15: variable 'B' is unbound
    %    6| also_wrong1(#{B := A} = #{b := B}) -> {A,B}.
    %     |               ^

    t.erl:7:27: variable 'A' is unbound
    %    7| also_wrong2(#{a := A} = #{A := B}) -> {A,B}.
    %     |                           ^

The other way to use `=` is in a function body. Used in this way,
the right-hand side must be an expression:

    Pattern = Expression

For example:

    foobar(Value) ->
        #{a := A} = #{a => Value},
        A.

Used in this context, the right-hand side of `=` must **not** be a pattern:

    illegal_foobar(Value) ->
        #{a := A} = #{a := Value},
        A.

    t.erl:18:21: only association operators '=>' are allowed in map construction
    %   18|     #{a := A} = #{a := Value},
    %     |                     ^

When used in a body context, the value of the `=` operator is the
value of its right-hand side operand. When multiple `=` operators are
combined, they are evaluted from right to left. That means that any
number of patterns can be matched at once:

    Pattern1 = Pattern2 = ... = PatternN = Expr

which is equivalent to:

    Var = Expr
    PatternN = Var
       .
       .
       .
    Pattern2 = Var
    Pattern1 = Var

Given that there is a well-defined evaluation order from right to
left, one would expect that the following example would be legal:

    baz(M) ->
        #{K := V} = #{k := K} = M,
        V.

It is not. In Erlang/OTP 25 or earlier, the compilation fails with the
following message:

    t.erl:28:7: variable 'K' is unbound
    %   28|     #{K := V} = #{k := K} = M,
    %     |       ^

That restriction is now lifted, making the example legal.

Closes #6348
Closes #6444
Closes #6467

RobinMorisset commented:
Thank you for this change! It makes the language a bit simpler, and I like that it clarifies the semantics of assignment.

RobinMorisset commented:
This PR seems to have introduced https://github.com/erlang/otp/issues/6445

bjorng commented:
@RobinMorisset Thanks for catching this problem. Fixed.

kikofernandez reviewed (approved):
Great work! I only have a minor suggestion to simplify a function implementation.

bjorng commented on lib/compiler/src/v3_core.erl:
```diff
@@ -2025,10 +2025,18 @@ is_safe(_) -> false.
-    {P1,E1} = fold_match(E0, P),
-    {{match,L,P0,P1},E1};
-fold_match(E, P) -> {P,E}.
+fold_match(P, E) ->
```
Thanks! That **is** simpler and works if you add a clause to handle the case that the first argument is not a `match` tuple:

```
fold_match({match, L0, P0, E1}, E0) ->
    fold_match_helper(E1, {match, L0, P0, E0});
fold_match(P, E) ->
    {E, P}.
```

Staring some more at the code, I saw that it could be further simplified by removing `fold_match/2` and renaming the helper function to `fold_match/2`. I have pushed a fixup commit that does that.

kikofernandez replied:
Perfect!

RobinMorisset commented:
This PR introduces an erlc crash on the following code:
```erlang
-module(validator).
-compile([export_all]).

f(#{[] := <<_>>, [] := <<_>>}) -> 
    ok.
``` 
It results in the following error message:
```
Function: f/1
minimized/validator.erl: internal error in pass beam_validator_strong:
exception error: no case clause matching {test_tail,8}
  in function  beam_validator:validate_bs_match/4 (beam_validator.erl, line 1743)
  in call from beam_validator:branch/4 (beam_validator.erl, line 2889)
  in call from beam_validator:validate_instrs/4 (beam_validator.erl, line 333)
  in call from beam_validator:validate_1/5 (beam_validator.erl, line 276)
  in call from beam_validator:validate_0/4 (beam_validator.erl, line 114)
  in call from beam_validator:validate/2 (beam_validator.erl, line 56)
  in call from compile:beam_validator_1/3 (compile.erl, line 1663)
  in call from compile:fold_comp/4 (compile.erl, line 408)
```
I could not reproduce this bug on master, only with this PR.

@bjorng I'm not sure whether such PR-specific bug reports should go here or in their own issue, please tell me if you have any preference (or more generally if I should report the bugs I find in a different way).

bjorng commented:
As long as a PR is not merged, I prefer bug reports to go into that PR as comments.

bjorng commented:
I have pushed a bug fix. Thanks for noticing the problem.

bjorng commented:
Thanks! Fixed.

RobinMorisset commented:
Here is a testcase that breaks with the latest version of this PR:
```erlang
f(X, <<Y>>) when ((Y > X) xor X); (not X); X ->
    ok.
```
Error message:
```
Sub pass sanitize
Function: f/2
minimized/precodegen.erl: internal error in pass beam_ssa_pre_codegen:
exception error: no match of right hand side value [{b_literal,integer},
 {b_var,13},
 {b_literal,[unsigned,big]},
 {b_literal,8},
 {b_literal,1}]
  in function  beam_ssa_pre_codegen:sanitize_instr/3 (beam_ssa_pre_codegen.erl, line 900)
  in call from beam_ssa_pre_codegen:do_sanitize_is/9 (beam_ssa_pre_codegen.erl, line 805)
  in call from beam_ssa_pre_codegen:sanitize/5 (beam_ssa_pre_codegen.erl, line 667)
  in call from beam_ssa_pre_codegen:sanitize/1 (beam_ssa_pre_codegen.erl, line 657)
  in call from compile:run_sub_passes_1/3 (compile.erl, line 422)
  in call from beam_ssa_pre_codegen:function/2 (beam_ssa_pre_codegen.erl, line 160)
  in call from beam_ssa_pre_codegen:functions/2 (beam_ssa_pre_codegen.erl, line 89)
  in call from beam_ssa_pre_codegen:module/2 (beam_ssa_pre_codegen.erl, line 84)
```

RobinMorisset commented:
And here is another with sensibly the same error message, I'm putting it here in case it helps with debugging:
```erlang
f(<<X>>, Y) when {(X / 1), (Y andalso true)}; Y ->
    ok.
```

RobinMorisset commented:
A third one with the same error message:
```erlang
f(<<X>>, Y) when ((<<X>> == false) orelse (Y andalso true)); Y ->
    Y.
```

bjorng commented:
Thanks! Fixed. I included all three examples in the test suite.

RobinMorisset commented:
Thank you for your quick fixes.
Here is a fresh testcase that crashes erlc with this PR but not without:
```erlang
f(X) ->
    case 0 of
        <<_:(X bor X)>> ->
            <<_:Y>> = (Y = X)
    end,
    Y.
```
The error message is:
```
Function: f/1
minimized/kernel_to_ssa6.erl: internal error in pass beam_kernel_to_ssa:
exception error: bad key: 'Y'
  in function  map_get/2
     called as map_get('Y',
 #{0 => {b_var,0},
   '@exception' => {b_var,'@exception'},
   '@ssa_bool' => {b_var,'@ssa_bool'},
   '@ssa_ret' => {b_var,'@ssa_ret'}})
     *** argument 1: not present in map
  in call from beam_kernel_to_ssa:ssa_arg/2 (beam_kernel_to_ssa.erl, line 1127)
  in call from beam_kernel_to_ssa:cg/2 (beam_kernel_to_ssa.erl, line 131)
  in call from beam_kernel_to_ssa:cg/2 (beam_kernel_to_ssa.erl, line 114)
  in call from beam_kernel_to_ssa:do_match_cg/3 (beam_kernel_to_ssa.erl, line 167)
  in call from beam_kernel_to_ssa:cg_fun/2 (beam_kernel_to_ssa.erl, line 91)
  in call from beam_kernel_to_ssa:function/3 (beam_kernel_to_ssa.erl, line 75)
  in call from beam_kernel_to_ssa:'-functions/3-lc$^0/1-0-'/3 (beam_kernel_to_ssa.erl, line 65)
```

RobinMorisset commented:
Another one that fails with this PR (both with and without 28fce1f8016a8bf52ca6a109aa471700f17e772b), and does not on master (both with and without 28fce1f8016a8bf52ca6a109aa471700f17e772b):
```erlang
f(#{0 := <<_:0>>, 0 := <<_>>}) -> 
    ok.
```
results in
```
Function: f/1
minimized/bug1.erl: internal error in pass beam_validator_strong:
exception error: no case clause matching {test_tail,0}
  in function  beam_validator:validate_bs_match/4 (beam_validator.erl, line 1716)
  in call from beam_validator:branch/4 (beam_validator.erl, line 2879)
  in call from beam_validator:validate_instrs/4 (beam_validator.erl, line 333)
  in call from beam_validator:validate_1/5 (beam_validator.erl, line 276)
  in call from beam_validator:validate_0/4 (beam_validator.erl, line 114)
  in call from beam_validator:validate/2 (beam_validator.erl, line 56)
  in call from compile:beam_validator_1/3 (compile.erl, line 1663)
  in call from compile:fold_comp/4 (compile.erl, line 408)
```

bjorng commented:
Thanks! Fixed.

RobinMorisset commented:
It seems to still be possible to trigger the same bug (or at least one with the same error message) with different testcases, such as:
```erlang
f(#{0 := <<_>>, 0 := <<X>>})  -> 
	X.
```
or
```erlang
f(#{0 := <<_>>, 0 := <<_, 0>>}) ->
    ok.
```
or
```erlang
f(#{0 := <<_>>, 0 := <<_, _:(0 div 0)>>}) ->
    0.
```
or
```erlang
f(#{0 := <<X:0>>, 0 := <<X>>}) ->
    ok.
```
or
```erlang
f() ->
    case (catch <<0 || true>>) of
        #{[] := <<_>>, [] := <<X>>} ->
            X
    end.
```

bjorng commented:
Yes, it is the same bug. I have now pushed a fix that cares of it in a different way than my previous attempts.

RobinMorisset commented:
Thank you for your reactivity. The fuzzer can no longer trigger that particular bug, but it now found some very similar testcases that trigger a different error message:
```erlang
f(#{0 := <<_, _:(true andalso 0)>>, 0 := <<_>>}) ->
    ok.
```
and
```erlang
f() ->
    case catch ({}+{}) of
        #{[] := <<X:0, _:X>>, [] := <<_>>} ->
            ok
    end.
```
Both cause a crash with the following error message (that is for the second case, the first one is the same except for some numbers):
```
Sub pass ssa_opt_bsm_shortcut
Function: f/0
minimized/bug2.erl: internal error in pass beam_ssa_opt:
exception error: bad key: {b_var,14}
  in function  map_get/2
     called as map_get({b_var,14},#{39 => {0,8},{b_var,13} => 0})
     *** argument 1: not present in map
  in call from beam_ssa_opt:bsm_positions/2 (beam_ssa_opt.erl, line 1959)
  in call from beam_ssa_opt:ssa_opt_bsm_shortcut/1 (beam_ssa_opt.erl, line 1944)
  in call from compile:run_sub_passes_1/3 (compile.erl, line 422)
  in call from beam_ssa_opt:phase/4 (beam_ssa_opt.erl, line 114)
  in call from beam_ssa_opt:run_phases/3 (beam_ssa_opt.erl, line 77)
  in call from beam_ssa_opt:module/2 (beam_ssa_opt.erl, line 69)
  in call from compile:'-select_passes/2-anonymous-0-'/3 (compile.erl, line 681)
```

bjorng commented:
Your latest example triggered a bug in the `ssa_opt_bsm` sub pass. I fixed it... I thought. The new test cases based on your examples fail when some optimizations are disabled. It seems to be an unrelated bug. I can't push my fix to this PR until I have fixed that bug, too, because there would be failed test cases in our daily builds. Meanwhile, if your want to find more bugs (with all compiler optimizations enabled), here is another branch which includes my latest fix:

https://github.com/bjorng/otp/tree/bjorn/compiler/parallel-match-WIP

bjorng commented:
I have now fixed the other bugs and updated this pull request.

bjorng commented:
@rvirding Do you have any opinions about this pull request and the `=` operator?

In particular, what do you think about that in bodies, multiple `=` operators will be evaluated from right to left? As an example, if the matching is defined to be right to left, that will make the following example legal:

```
baz(M) ->
    #{K := V} = #{k := K} = M,
    V.
```

I am asking because we have had some discussions in the OTP team as a preparation for an OTP Technical Board meeting we will have later this week.

bjorng commented:
The OTP Technical Board has now approved this pull request.

 I have updated the documentation to clarify the semantics of the **match operator** and the **compound pattern operator** (a new name invented by @RaimoNiskanen in order to more easily distinguish the two flavors of the `=` operator).

After the updates of the documentation have been reviewed, I hope to merge this pull request next week.

RobinMorisset commented on system/doc/reference_manual/expressions.xml:
```diff
@@ -192,22 +201,112 @@ case {Value, Result} of
+2> <input>F = fun({A, B} = E) -> {E, A + B} end, F({1,2}).</input>
+{{1,2},3}
+3> <input>G = fun(&lt;&lt;A:8,B:8>> = &lt;&lt;C:16>>) -> {A, B, C} end, G(&lt;&lt;42,43>>).</input>
+{42,42,10795}</pre>
```
Shouldn't the second element be 43 here?

bjorng replied:
Yes, it should. I changed example to make it clearer and changed it manually instead of copy-pasting the whole thing from the output of the shell.

RobinMorisset commented on system/doc/reference_manual/expressions.xml:
```diff
@@ -192,22 +201,112 @@ case {Value, Result} of
     <p>If the matching succeeds, any unbound variable in the pattern
-      becomes bound and the value of <c>Expr2</c> is returned.</p>
+    becomes bound and the value of <c>Expr</c> is returned.</p>
+    <p>If multiple match operators are applied in sequence, they will be
```
I'd maybe add that even if there is a single match operator, its right hand side will be executed before the pattern on its left hand side is checked. So for example the following is valid:
```erlang
<<X:Y>> = begin Y = 8, <<42:8>> end
```
And it correctly binds X to 42, as Y is bound to 8 before the validity of the pattern on the left hand side is checked.

bjorng replied:
Yes, I will add that clarification and an example.

RobinMorisset reviewed (commented):
Thank you again for your work on this!

RaimoNiskanen reviewed (approved):
The documentation changes looks very good to me

rickard-green reviewed (approved):
The doc changes looks good to me too :+1:
